### PR TITLE
Add file_row_number flag to parquet reader

### PR DESCRIPTION
**Pull request by hannes**
for unspecified titanic purposes

### CODE REVIEW

File: extension/parquet/include/row_number_column_reader.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#ifndef DUCKDB_AMALGAMATION
+#include "duckdb/common/limits.hpp"
+#endif
+#include "column_reader.hpp"
+#include "templated_column_reader.hpp"
+
+namespace duckdb {
+
+//! Reads a constant that is not read from the parquet file but instead generated by the parquet reader, used for
+//! filename column. Can be used for handling hive partitioning later.
```
Comment (samansmink) [inline extension/parquet/include/row_number_column_reader.hpp:20]: nit: comment

File: test/sql/copy/parquet/parquet_row_number.test [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# name: test/sql/copy/parquet/parquet_row_number.test
+# description: Test the filename option of the parquet reader
```
Comment (samansmink) [inline test/sql/copy/parquet/parquet_row_number.test:2]: nit: row number

File: extension/parquet/include/row_number_column_reader.hpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+class RowNumberColumnReader : public ColumnReader {
+public:
+	RowNumberColumnReader(ParquetReader &reader, LogicalType type_p, const SchemaElement &schema_p, idx_t schema_idx_p,
+	                      idx_t max_define_p, idx_t max_repeat_p);
+
+public:
+	idx_t Read(uint64_t num_values, parquet_filter_t &filter, uint8_t *define_out, uint8_t *repeat_out,
+	           Vector &result) override;
+
+	unique_ptr<BaseStatistics> Stats(const std::vector<ColumnChunk> &columns) override {
```
Comment (samansmink) [inline extension/parquet/include/row_number_column_reader.hpp:30]: Can we maybe add the stats here? this allows proper handling of pushed down filters on this column

File: test/sql/copy/parquet/parquet_row_number.test [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# group: [parquet]
+
+require parquet
+
+query III
+select min(file_row_number), max(file_row_number), avg(file_row_number) from parquet_scan('data/parquet-testing/lineitem-top10000.gzip.parquet', file_row_number=1, filename=1);
```
Comment (samansmink) [inline test/sql/copy/parquet/parquet_row_number.test:8]: Could we test here with a filter by row number?

**Review by samansmink — COMMENTED** [review]
added a few comments!

File: test/sql/copy/parquet/parquet_row_number.test (hunk 4 above)
Comment (hannes) [reply test/sql/copy/parquet/parquet_row_number.test]: added this, also added a test where the first couple of row groups are eliminated using parquet stats, then the file_row_number also needs to produce the correct result

**Comment by hannes** [comment]
@samansmink added the per-row group stats, have a look

**Comment by samansmink** [comment]
yea looks good!

**Pull request merged**